### PR TITLE
Drop U16 and add U20 support

### PR DESCRIPTION
**Pull request by winem**
This PR drops the support for Ubuntu 16 (xenial) and adds support for Ubuntu 20 (focal). 

In addition this PR also adds meta/main.yml files for the redis and smoketests role which weren't there yet. This is rather a matter of consistency than of functionality but it seems a good chance to get this done.

It also removes support for MongoDB-Versions that are already EoL as there are 3.2 and 3.4.

Tests with st2repo_name = unstable were successful.

### CODE REVIEW

**Comment by armab** [comment]
Looks like we'll also need to add U 20 to CI, per https://travis-ci.org/github/StackStorm/ansible-st2/builds/773942784

File: roles/StackStorm.redis/meta/main.yml
Comment (cognifloyd) [inline roles/StackStorm.redis/meta/main.yml:27]: Why does the `StackStorm.redis` role need the st2repo?

File: roles/StackStorm.st2/tasks/auth.yml
Comment (cognifloyd) [inline roles/StackStorm.st2/tasks/auth.yml:5]:
```suggestion
    - "{{ ansible_facts.os_family | lower }}_{{ ansible_facts.distribution_major_version }}.yml"
    - "{{ ansible_facts.os_family | lower }}.yml"
```

File: roles/StackStorm.mongodb/vars/debian_20.yml
Comment (cognifloyd) [inline roles/StackStorm.mongodb/vars/debian_20.yml:4]: Duplicating these keys is not ideal. Is the idea here to prevent people from trying to install 3.6 or 4.0 on debian 20?

**Review by cognifloyd — CHANGES_REQUESTED** [review]
I have few minor changes and questions.

File: roles/StackStorm.mongodb/vars/debian.yml
Comment (amanda11) [inline roles/StackStorm.mongodb/vars/debian.yml:5]: I think as we haven't yet tested with 4.2, then I'd be tempted to only add 4.4 for now.
Add 4.2 when we confirm if we want to support it as part of 3.6

**Review by amanda11 — CHANGES_REQUESTED** [review]
As well as the comment on mongo 4.2, I think we also need updates on:

- .kitchen.yml - has ubuntu 16 still
- .travis.yml
-

File: roles/StackStorm.mongodb/vars/debian_20.yml
Comment (winem) [reply roles/StackStorm.mongodb/vars/debian_20.yml:4]: I don't think that it would prevent people from installing an older version. There is no package for mongodb 3.6 or 4.0 on focal. So trying to install mongodb 4.0 for example would fail while adding the key as such would instead work. 

So the keys aren't needed for U20 and I thought duplication is acceptable as it keeps the code clean and the task to retrieve the variables does not need to be changed: 
```
---
- name: Retrieve variables
  include_vars: "{{ item }}"
  with_first_found:
    - "{{ ansible_os_family | lower }}_{{ ansible_distribution_major_version }}.yml"
    - "{{ ansible_os_family | lower }}.yml"
  tags: [databases, mongodb]
``` 

I think I'm going to add another task that loads the debian.yml that has all the apt keys and the debian_20.yml only if it's a U20 setup to override the mongodb version. Will push it after a test.

Comment (winem) [reply roles/StackStorm.mongodb/vars/debian_20.yml]: Done with https://github.com/StackStorm/ansible-st2/pull/298/commits/df3e85cedcca5a63f94dddcfac8cc28b89d0652d

File: roles/StackStorm.redis/meta/main.yml
Comment (winem) [reply roles/StackStorm.redis/meta/main.yml:27]: You are right, there is no technical dependency between both. Removed it from the meta file now.

File: roles/StackStorm.mongodb/vars/debian.yml
Comment (winem) [reply roles/StackStorm.mongodb/vars/debian.yml:5]: Ok, fine for me. I removed it.

**Comment by winem** [comment]
Will update the ci files (travis, kitchen, etc.) till tomorrow evening.

**Comment by cognifloyd** [comment]
`mongodb_version` needs to change for debian 20. You added an extra vars file and tasks to load that. I think it would be cleaner to consolidate the version choice in the defaults file, which looks like:

https://github.com/StackStorm/ansible-st2/blob/94e3c4f37e023b2783b0de4294f2dc154d73150f/roles/StackStorm.mongodb/defaults/main.yml#L2-L3

But you could do this:
```diff
 # MongoDB default version
-mongodb_version: "4.0" 
+mongodb_version: "{% if ansible_facts.os_family == 'Debian' and ansible_facts.distribution_major_version == '20' %}4.4{% else %}4.0{% endif %}" 
```

File: roles/StackStorm.mongodb/tasks/main.yml
Comment (armab) [inline roles/StackStorm.mongodb/tasks/main.yml:18]: This 3-tasks construct doesn't look good indeed.
Avoiding os-specific `when` condition is exactly the purpose of the 1st task `with_first_found` which is a known best practice.

Comment (winem) [reply roles/StackStorm.mongodb/tasks/main.yml]: That's true and imho it's actually an anti-pattern to override defaults via role variables. This has a high potential to cause failures and confuse maintainers. Now I switched to the solution suggested by @cognifloyd which seems to be the most elegant one.

**Comment by winem** [comment]
> `mongodb_version` needs to change for debian 20. You added an extra vars file and tasks to load that. I think it would be cleaner to consolidate the version choice in the defaults file, which looks like:
> 
> https://github.com/StackStorm/ansible-st2/blob/94e3c4f37e023b2783b0de4294f2dc154d73150f/roles/StackStorm.mongodb/defaults/main.yml#L2-L3
> 
> But you could do this:
> 
> ```diff
>  # MongoDB default version
> -mongodb_version: "4.0" 
> +mongodb_version: "{% if ansible_facts.os_family == 'Debian' and ansible_facts.distribution_major_version == '20' %}4.4{% else %}4.0{% endif %}" 
> ```

I did it like that now. Guess that's the smartest solution to the issue and does not break with any common pattern or best practices.

**Review by amanda11 — APPROVED** [review]
LGTM - but wait for those others that gave review comments to give their approval.

**Review by cognifloyd — APPROVED** [review]
LGTM

File: .kitchen.yml
Comment (armab) [inline .kitchen.yml:36]: You probably meant `focal` here:
```suggestion
      image: stackstorm/packagingtest:focal-systemd
```

File: .travis.yml
Comment (amanda11) [inline .travis.yml:13]: Need to remove ubuntu-20 from here. We can't run it as a stable run yet until 3.5 is released. 
So we need to do the changes for ansible-st2 in 2 stages.
1) add it to unstable before release - under this PR
2) add it to stable after release - separate PR

**Review by amanda11 — CHANGES_REQUESTED** [review]

File: .travis.yml
Comment (armab) [reply .travis.yml:13]: I think we'll just have to wait for a release to happen before merging this PR.

Otherwise, someone using the fresh Ansible installer will have the same issues with U16/U20.

File: roles/StackStorm.mongodb/defaults/main.yml
Comment (rush-skills) [inline roles/StackStorm.mongodb/defaults/main.yml:3]: To decide between 4.0 and 4.4 mongo version, wouldn't it be a better idea to use st2 version > 3.4 rather than only on Ubuntu 20?

Comment (amanda11) [reply roles/StackStorm.mongodb/defaults/main.yml:3]: @rushskills Stackstorm 3.5 still uses mongo 4.0 for all OSes EXCEPT for Ubuntu 30

**Review by armab — APPROVED** [review]
:+1: 
Following the #Ansible community messages, we want to merge it, considering the 3.5.0 release is ready.

Just migrated repo to the new TravisCI.com to get the build finished.

**Comment by armab** [comment]
All green, great work everyone!

**Pull request merged**
